src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerBase.java

```diff
@@ -0,0 +1,144 @@
+ * {@link #onTraversalSuccess} and/or {@link #closeTool}.
+ */
+public abstract class VariantWalkerBase extends GATKTool {
+
+    /**
+     * This number controls the size of the cache for our primary and auxiliary FeatureInputs
+     * (specifically, the number of additional bases worth of overlapping records to cache when querying feature sources).
+     */
+    public static final int FEATURE_CACHE_LOOKAHEAD = 100_000;
+
+    @Override
+    public boolean requiresFeatures() { return true; }
+
+    @Override
+    void initializeFeatures() {
+        //Note: we override this method because we don't want to set feature manager to null if there are no FeatureInputs.
+        //This is because we have at least 1 source of features (namely the driving dataset).
+        features = new FeatureManager(this, FEATURE_CACHE_LOOKAHEAD);
+        initializeDrivingVariants();
+    }
+
+    /**
+     * Overriding the superclass method to preferentially
+     * choose the sequence dictionary from the driving source of variants.
+     */
+    @Override
+    public final SAMSequenceDictionary getBestAvailableSequenceDictionary() {
+        final SAMSequenceDictionary dictFromDrivingVariants = getSequenceDictionaryForDrivingVariants();
+        if (dictFromDrivingVariants != null) {
+            //If this dictionary looks like it was synthesized from a feature index, see
+            //if there is a better dictionary available from another source (i.e., the reference)
+            if (IndexUtils.isSequenceDictionaryFromIndex(dictFromDrivingVariants)) {
+                final SAMSequenceDictionary otherDictionary = super.getBestAvailableSequenceDictionary();
+                return otherDictionary != null ?
+                        otherDictionary :
+                        dictFromDrivingVariants;
+            } else {
+                return dictFromDrivingVariants;
+            }
+        }
+        return super.getBestAvailableSequenceDictionary();
+    }
+
+    /**
+     * Process the feature inputs that represent the primary driving source(s) of variants for this tool, and
+     * perform any necessary header and sequence dictionary validation. Called by the framework during feature
+     * initialization.
+     */
+    protected abstract void initializeDrivingVariants();
+
+    /**
+     * Return the VCFHeader to be used for the driving variants for this tool. The value returned will usually
+     * have been prepared in {@link #initializeDrivingVariants}
+     * @return VCFHeader to be used for the driving variants.
+     */
+    public abstract VCFHeader getHeaderForVariants();
+
+    /**
+     * Return the primary sequence dictionary to be used for the driving variants for this tool. The value returned
+     * will usually have been prepared in {@link #initializeDrivingVariants}
+     */
+    protected abstract SAMSequenceDictionary getSequenceDictionaryForDrivingVariants();
```

Comment:
I take it back. It's always the same for a MultiVariantWalker, because we make it so by ensuring the header contains the merged dictionary. But for a VariantWalker with a single input with no embedded dictionary, but with an accompanying index, the header will not have the dictionary, but FeatureDataSource will return the synthesized one . Maybe we should make FeatureDataSource behave like MultiFeatureDataSource (over the TODO above) ? Also, I’m wondering if it better perf-wise to cache the dictionary once rather than regenerating it every time from the header, given that hg38 apparently has thousands of contigs.